testsuite/model/src/test/java/org/keycloak/testsuite/model/transaction/StorageTransactionTest.java

```diff
@@ -0,0 +1,185 @@
+        TransactionController tx3 = new TransactionController(getFactory());
+
+        tx1.begin();
+        tx2.begin();
+        tx3.begin();
+
+        // Read original value in tx1
+        assertThat(
+                tx1.runStep(session -> session.realms().getRealm(realmId).getAttribute("k1")),
+                equalTo("v1"));
+
+        // change value to new in tx2
+        tx2.runStep(session -> {
+            session.realms().getRealm(realmId).setAttribute("k1", "v2");
+            return null;
+        });
+        tx2.commit();
+
+        // tx1 should still return the value that already read
+        assertThat(
+                tx1.runStep(session -> session.realms().getRealm(realmId).getAttribute("k1")),
+                equalTo("v1"));
+
+        // tx3 should return the new value
+        assertThat(
+                tx3.runStep(session -> session.realms().getRealm(realmId).getAttribute("k1")),
+                equalTo("v2"));
+        tx1.commit();
+        tx3.commit();
+
+    }
+
+    @Test
+    // LockObjectForModification is currently used only in map-jpa
+    @RequireProvider(value = MapStorageProvider.class, only = JpaMapStorageProviderFactory.PROVIDER_ID)
+    public void testLockObjectForModification() {
+        CONFIG.spi(MapStorageSpi.NAME)
+                .provider(JpaMapStorageProviderFactory.PROVIDER_ID)
+                .config("lockTimeout", "1000");
+
+        TransactionController tx1 = new TransactionController(getFactory());
+        TransactionController tx2 = new TransactionController(getFactory());
+        TransactionController tx3 = new TransactionController(getFactory());
+
+        tx1.begin();
+        tx2.begin();
+
+        // tx1 acquires lock
+        tx1.runStep(session -> LockObjectsForModification.lockRealmsForModification(session, () -> session.realms().getRealm(realmId)));
+        try {
+            // tx2 should fail as tx1 is holding the lock
+            tx2.runStep(session -> LockObjectsForModification.lockRealmsForModification(session, () -> session.realms().getRealm(realmId)));
+            throw new AssertionError("Should not succeed as tx1 locked the realm");
+        } catch (RuntimeException ex) {
+            // correct
+        }
+        // end both transactions
+        tx2.rollback();
+        tx1.commit();
+
+        // start new transaction and read again, it should be successful
+        tx3.begin();
+        tx3.runStep(session -> LockObjectsForModification.lockRealmsForModification(session, () -> session.realms().getRealm(realmId)));
+        tx3.commit();
+    }
+
+    @Test
+    public void testOptimisticLockingException() {
+        // This test doesn't work with hotrod, map-ldap and map-chm
+        String realmMapStorageProvider = CONFIG.getConfig().get("realm.map.storage.provider");
+        assumeTrue("Optimistic locking works only with the legacy store and map-jpa",
+                realmMapStorageProvider == null || "jpa".equals(realmMapStorageProvider));
```

Comment:
The tests fail for the profile `map` which has the ConcurrentHashMap. I assume this condition needs a refinement.

Reply:
should be fixed now